File: Backtracking_and_Search/LC_40_Combination_sum_II.py
# Written by: Jian-Yuan Lin
# Date: 2022.11.19

class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        candidates.sort()
        ans = []
        def dfs(nums, t, start, curr, ans):
            if t == 0:
                ans.append(curr[:])
                return
            if t < 0:
                return
            for i in range(start, len(nums)):
                if i != start and nums[i] == nums[i-1]: continue
                curr.append(nums[i])
                dfs(nums, t - nums[i], i+1, curr, ans)
                curr.pop()
                
        dfs(candidates, target, 0, [], ans)

        return ans

File: Backtracking_and_Search/LC_40_Combination_sum_II.cpp
// Written by: Jian-Yuan Lin
// Date: 2022.11.19

class Solution {
public:
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        sort(candidates.begin(), candidates.end());
        vector<vector<int>> ans;
        vector<int> curr;
        dfs(candidates, target, 0, curr, ans);
        return ans;
    }
private:
    void dfs(vector<int>& nums, int target, int start, vector<int>& curr, vector<vector<int>>& ans){
        if (target == 0){
            ans.push_back(curr);
            return;
        }
        if (target < 0){
            return;
        }
        for (int i = start; i < nums.size(); i++){
            if (i != start && nums[i] == nums[i-1]){
                continue;
            }
            curr.push_back(nums[i]);
            dfs(nums, target - nums[i], i+1, curr, ans);
            curr.pop_back();
        }
    }
};